commandapi-core/src/main/java/dev/jorel/commandapi/CommandAPIHandler.java

```diff
@@ -615,15 +618,22 @@ void register(CommandMetaData<CommandSender> meta, final Argument[] args,
 			argumentsString.add(arg.getNodeName() + ":" + arg.getClass().getSimpleName());
 		}
 		RegisteredCommand registeredCommandInformation = new RegisteredCommand(commandName, argumentsString, shortDescription,
-			fullDescription, usageDescription, aliases, permission);
+			fullDescription, usageDescription, aliases, permission, namespace);
+		registeredCommands.add(registeredCommandInformation);
+
+		List<RegisteredCommand> registeredCommands = registeredCommandMap.containsKey(commandName)
+			? registeredCommandMap.get(commandName)
+			: new ArrayList<>();
 		registeredCommands.add(registeredCommandInformation);
+		registeredCommandMap.put(commandName, registeredCommands);
 
 		// Handle previewable arguments
 		handlePreviewableArguments(commandName, args, aliases);
 
 		platform.preCommandRegistration(commandName);
 
-		CommandAPI.logInfo("Registering command /" + commandName + " " + humanReadableCommandArgSyntax);
+		String namespacedCommandName = namespace.isEmpty() ? commandName : namespace +  ":" + commandName;
```

Comment:
It's possible I've missed something, but is the only case when `namespace.isEmpty()` will occur be for Velocity?

Reply:
Nope, there's one instance in `CommandAPIVelocity#registerCommandNode(LiteralCommandNode, String)`

Reply:
> Nope, there's one instance in `CommandAPIVelocity#registerCommandNode(LiteralCommandNode, String)`

In other words, will `namespace.isEmpty()` always be `false` when running on a Bukkit server?

Reply:
Actually, no. It is perfectly valid to register a command using a `""` String literal in the register(String) method

Reply:
Should we be allowing a user to use `""` as a namespace in the `register(String)` method for Bukkit?

Reply:
I actually don't know and had the same concern. @willkroboth [mentioned](https://github.com/JorelAli/CommandAPI/pull/509#discussion_r1418932864) something about that here which is why I allowed that.

Reply:
My counter-example to "a command with no namespace" is simply the fact that in the Bukkit command system, _all commands should have namespaces_. If I have two commands `minecraft:reload` and `bukkit:reload`, if I run `/reload`, it will choose the command with the highest "priority" according to however Bukkit assigns priority. In Bukkit's ecosystem, commands with "no namespace" are automatically given the namespace of their registering plugin.

From a Bukkit point of view, I don't think we should be allowing commands with no namespaces. While _yes_ they will probably all work out fine, semantically it's getting a little bit non-Bukkit-ty and I don't like the idea of that.

Reply:
I do agree with that. I absolutely do.
Another point I was thinking about were the currently existing exceptions. I originally added those really early when I still had the `@Nonnull` annotation on parameters.
I believe we could change this behaviour to instead log a message, maybe only if verbose outputs are true or something and register the command under the default namespace instead.
Or do we keep *an* exception that doesn't necessarily is a NPE.
Depending on this change we should be implementing a similar behaviour when an empty String is passed into that method.

Reply:
> If I have two commands minecraft:reload and bukkit:reload, if I run /reload, it will choose the command with the highest "priority" according to however Bukkit assigns priority.

Yeah, that makes sense. A note on the priority system: Bukkit chooses whichever command is registered first. I think it would be good if we complied with this on the following line, by only adding our command to the map if there isn't already a key associated with the command name.

https://github.com/JorelAli/CommandAPI/blob/d37d424072d6381639d2027f2ab32bcabd0464aa/commandapi-platforms/commandapi-bukkit/commandapi-bukkit-core/src/main/java/dev/jorel/commandapi/CommandAPIBukkit.java#L561

Imo, since Velocity has no-namespace commands as the default behavior, it makes sense for the API design to allow that in Bukkit as well. I think it's important to allow the developer to choose what they want to do. `CommandAPIHandler` can internally already handle an empty namespace for Velocity, so we can also just expose that to the developer in Bukkit as well.

Reply:
> Imo, since Velocity has no-namespace commands as the default behavior, it makes sense for the API design to allow that in Bukkit as well.

I am not so sure I agree with this. No-namespace commands are a Velocity feature already, on Bukkit they are not. We should follow this system - even if we could do different.
For me this means that we could allow no-namespace commands for Velocity since they are already a thing but we don't allow no-namespace commands on Bukkit because they are not a thing there.

Reply:
🤷 Alright, I suppose it helps simplify some logic.

For example, in these lines:

https://github.com/JorelAli/CommandAPI/blob/d37d424072d6381639d2027f2ab32bcabd0464aa/commandapi-platforms/commandapi-bukkit/commandapi-bukkit-core/src/main/java/dev/jorel/commandapi/CommandAPIBukkit.java#L557-L574

Instead of using reflection to access the `knownCommands` map, we can just use [`CommandMap#register(String fallbackPrefix, Command command)`](https://hub.spigotmc.org/javadocs/bukkit/org/bukkit/command/CommandMap.html#register(java.lang.String,org.bukkit.command.Command)) since the namespace will no longer be an empty String.

Reply:
Do we want to cancel the command registration with an exception or a message here?
If a message, do we want to log that even if verbose outputs are false?

Reply:
Let's use a message. We would probably want this to log despite verbose outputs set to false.

If the namespace is null, we should NPE.

If the namespace is empty, could we fall back to the `minecraft` for Bukkit (and if so, add that to the message as well?)

Reply:
Yeah, we would definitely fall back to the default namespace. I'd also be inclined to not throw a NPE and instead just log this as well and register using the default namespace. What do we think abou that?

Reply:
There should never be a case where a user registers a command with a "null" namespace - a user should not be passing null to CommandAPI methods. It is probably worth adding an explicit `@Nonnull` annotation in this case. If a user wants "null" as a namespace, they can just use the method that doesn't accept a namespace.

Reply:
So? I am not sure how to continue now:
- Throwing a NPE if a namespace was unexpectedly null
- Logging a message if a namespace was unexpectedly null
- Adding a `@Nonnull` annotation and still do one of the methods above because that annotation doesn't prevent using null, but instead only displays a warning

Reply:
If `null`, we throw an NPE (we don't log a message, we just throw). The only possible way for the namespace to be `null` is because a user has passed `null` to the namespace setting function, and we don't support that because users shouldn't be passing `null` to the CommandAPI.

We _could_ also add `@Nonnull`, but I don't think it's strictly necessary.